File: services/tag_reader.py
import os
import re
from pathlib import Path
from typing import Optional, Dict, Any, List
import mutagen
from mutagen.id3 import ID3
from mutagen.mp3 import MP3
from mutagen.flac import FLAC
from mutagen.mp4 import MP4
from mutagen.oggvorbis import OggVorbis

SUPPORTED_EXTENSIONS = {".mp3", ".flac", ".m4a", ".ogg", ".opus", ".wav"}

def clean_tag(val):
    if val is None:
        return None
    if isinstance(val, list):
        val = val[0] if val else None
    if val is None:
        return None
    s = str(val).strip()
    return s if s else None

def parse_year(date_str: Optional[str]) -> Optional[int]:
    if not date_str:
        return None
    match = re.search(r"\b(19\d{2}|20\d{2})\b", str(date_str))
    if match:
        return int(match.group(1))
    return None

def compute_decade(year: Optional[int]) -> Optional[str]:
    if not year or year < 1900 or year > 2099:
        return None
    return f"{(year // 10) * 10}s"

def parse_filename_fallback(filename: str):
    """
    Tries to extract Artist and Title from filename if tags are empty.
    e.g. 'Maddix - Acid For Breakfast.mp3' -> ('Maddix', 'Acid For Breakfast')
    """
    stem = Path(filename).stem
    # Remove track numbers like '01 - ' or '01. '
    stem = re.sub(r"^\d+[\s\.\-_]+", "", stem)
    
    if " - " in stem:
        parts = stem.split(" - ", 1)
        artist = parts[0].strip()
        title = parts[1].strip()
        return artist, title
    return None, stem

def read_audio_file(file_path: str) -> Optional[Dict[str, Any]]:
    path = Path(file_path)
    if not path.exists() or path.suffix.lower() not in SUPPORTED_EXTENSIONS:
        return None
    
    ext = path.suffix.lower()
    file_size = path.stat().st_size
    duration = 0.0
    
    title = None
    artist = None
    album = None
    year = None
    parent_genre = None
    genre = None
    sub_genre = None
    tempo = None
    tags_list = []
    has_cover = 0
    
    try:
        audio = mutagen.File(str(path))
        if audio is not None:
            if hasattr(audio, "info") and hasattr(audio.info, "length"):
                duration = round(audio.info.length, 1)
                
            if ext == ".mp3":
                # Handle MP3 ID3 tags
                try:
                    id3 = ID3(str(path))
                    title = clean_tag(id3.get("TIT2"))
                    artist = clean_tag(id3.get("TPE1"))
                    album = clean_tag(id3.get("TALB"))
                    
                    year_raw = clean_tag(id3.get("TDRC") or id3.get("TYER"))
                    year = parse_year(year_raw)
                    
                    # Custom TXXX frames
                    parent_txxx = id3.get("TXXX:PARENT_GENRE")
                    if parent_txxx:
                        parent_genre = clean_tag(parent_txxx.text)
                        
                    genre_txxx = id3.get("TXXX:GENRE")
                    if genre_txxx:
                        genre = clean_tag(genre_txxx.text)
                    else:
                        tcon_raw = clean_tag(id3.get("TCON"))
                        if tcon_raw:
                            # TCON might contain combined genre e.g. "Rock; Alternative Rock; Grunge"
                            if ";" in tcon_raw:
                                parts = [p.strip() for p in tcon_raw.split(";") if p.strip()]
                                if len(parts) >= 3:
                                    if not parent_genre: parent_genre = parts[0]
                                    genre = parts[1]
                                    if not sub_genre: sub_genre = parts[2]
                                elif len(parts) == 2:
                                    if not parent_genre: parent_genre = parts[0]
                                    genre = parts[1]
                                else:
                                    genre = parts[0]
                            else:
                                genre = tcon_raw
                        
                    sub_txxx = id3.get("TXXX:SUBGENRE")
                    if sub_txxx:
                        sub_genre = clean_tag(sub_txxx.text)
                        
                    # Tempo / Pace
                    tempo_txxx = id3.get("TXXX:TEMPO") or id3.get("TXXX:PACE")
                    if tempo_txxx:
                        tempo = clean_tag(tempo_txxx.text)
                    elif id3.get("TBPM"):
                        tempo = f"~{clean_tag(id3.get('TBPM'))} BPM"
                        
                    # Check cover
                    for k in id3.keys():
                        if k.startswith("APIC"):
                            has_cover = 1
                            break
                except Exception:
                    pass
                    
            elif ext == ".flac":
                try:
                    flac = FLAC(str(path))
                    title = clean_tag(flac.get("title"))
                    artist = clean_tag(flac.get("artist"))
                    album = clean_tag(flac.get("album"))
                    year = parse_year(clean_tag(flac.get("date") or flac.get("year")))
                    parent_genre = clean_tag(flac.get("parent_genre"))
                    raw_genre = clean_tag(flac.get("genre"))
                    if raw_genre and ";" in raw_genre:
                        parts = [p.strip() for p in raw_genre.split(";") if p.strip()]
                        if len(parts) >= 2 and not parent_genre:
                            parent_genre = parts[0]
                            genre = parts[1]
                        else:
                            genre = parts[0]
                    else:
                        genre = raw_genre
                    sub_genre = clean_tag(flac.get("subgenre") or flac.get("sub_genre"))
                    tempo = clean_tag(flac.get("tempo") or flac.get("pace"))
                    if not tempo and flac.get("bpm"):
                        tempo = f"~{clean_tag(flac.get('bpm'))} BPM"
                    if flac.pictures:
                        has_cover = 1
                except Exception:
                    pass
                    
            elif ext in [".m4a", ".mp4"]:
                try:
                    mp4 = MP4(str(path))
                    title = clean_tag(mp4.get("\xa9nam"))
                    artist = clean_tag(mp4.get("\xa9ART"))
                    album = clean_tag(mp4.get("\xa9alb"))
                    year = parse_year(clean_tag(mp4.get("\xa9day")))
                    raw_genre = clean_tag(mp4.get("\xa9gen"))
                    if raw_genre and ";" in raw_genre:
                        parts = [p.strip() for p in raw_genre.split(";") if p.strip()]
                        if len(parts) >= 2:
                            parent_genre = parts[0]
                            genre = parts[1]
                        else:
                            genre = parts[0]
                    else:
                        genre = raw_genre
                        
                    # Read custom iTunes parent genre and subgenre
                    if "----:com.apple.iTunes:PARENT_GENRE" in mp4:
                        raw = mp4["----:com.apple.iTunes:PARENT_GENRE"]
                        if raw:
                            val = raw[0].decode("utf-8", errors="ignore") if isinstance(raw[0], (bytes, bytearray)) else str(raw[0])
                            parent_genre = clean_tag(val)
                    if "----:com.apple.iTunes:SUBGENRE" in mp4:
                        raw = mp4["----:com.apple.iTunes:SUBGENRE"]
                        if raw:
                            val = raw[0].decode("utf-8", errors="ignore") if isinstance(raw[0], (bytes, bytearray)) else str(raw[0])
                            sub_genre = clean_tag(val)
                            
                    if "----:com.apple.iTunes:TEMPO" in mp4:
                        tempo = clean_tag(mp4["----:com.apple.iTunes:TEMPO"])
                    elif "tmpo" in mp4 and mp4["tmpo"]:
                        tempo = f"~{mp4['tmpo'][0]} BPM"
                    if "covr" in mp4 and mp4["covr"]:
                        has_cover = 1
                except Exception:
                    pass
    except Exception as e:
        print(f"Error reading metadata for {path.name}: {e}")
        
    # If title or artist missing, fallback to filename pattern
    if not title or not artist:
        fb_artist, fb_title = parse_filename_fallback(path.name)
        if not title:
            title = fb_title
        if not artist:
            artist = fb_artist
            
    decade = compute_decade(year)
    
    # Format tags list based on tags actually in file (do not fabricate parent tag; let Gemini decide!)
    from services.taxonomy import format_track_tags
    tags_list = format_track_tags(parent_genre, genre, sub_genre, decade)
        
    # Initial status:
    # If the file already has a parent_genre tag saved, mark 'saved'.
    # Otherwise mark 'unprocessed' so user knows it needs Gemini AI analysis!
    status = "saved" if parent_genre else "unprocessed"
    
    return {
        "file_path": str(path.resolve()),
        "file_name": path.name,
        "file_size": file_size,
        "duration": duration,
        "title": title or path.stem,
        "artist": artist or "Unknown Artist",
        "album": album or "",
        "release_year": year,
        "decade": decade,
        "parent_genre": parent_genre,
        "genre": genre,
        "sub_genre": sub_genre,
        "tags": tags_list,
        "status": status,
        "has_cover": has_cover
    }

def scan_directory(directory_path: str) -> List[Dict[str, Any]]:
    p = Path(directory_path)
    if not p.exists() or not p.is_dir():
        return []
        
    results = []
    for item in p.rglob("*"):
        if item.is_file() and item.suffix.lower() in SUPPORTED_EXTENSIONS:
            track = read_audio_file(str(item))
            if track:
                results.append(track)
    return results

def extract_cover_bytes(file_path: str) -> Optional[tuple[bytes, str]]:
    """
    Extracts raw album art bytes and mime type for in-browser display.
    """
    path = Path(file_path)
    if not path.exists():
        return None
        
    ext = path.suffix.lower()
    try:
        if ext == ".mp3":
            id3 = ID3(str(path))
            for k, v in id3.items():
                if k.startswith("APIC"):
                    return v.data, v.mime
        elif ext == ".flac":
            flac = FLAC(str(path))
            if flac.pictures:
                pic = flac.pictures[0]
                return pic.data, pic.mime
        elif ext in [".m4a", ".mp4"]:
            mp4 = MP4(str(path))
            if "covr" in mp4 and mp4["covr"]:
                data = bytes(mp4["covr"][0])
                # Check format
                mime = "image/png" if data.startswith(b"\x89PNG") else "image/jpeg"
                return data, mime
    except Exception as e:
        print(f"Error extracting cover from {path.name}: {e}")
        
    return None

File: server/tag_reader.js
const fs = require('fs');
const path = require('path');
const mm = require('music-metadata');
const { formatTrackTags, inferParentFromGenre } = require('./taxonomy');

const SUPPORTED_EXTENSIONS = new Set([".mp3", ".flac", ".m4a", ".ogg", ".opus", ".wav"]);

function cleanTag(val) {
  if (val === null || val === undefined) return null;
  if (Array.isArray(val)) {
    val = val.length > 0 ? val[0] : null;
  }
  if (val === null || val === undefined) return null;
  const s = String(val).trim();
  return s.length > 0 ? s : null;
}

function parseYear(dateStr) {
  if (!dateStr) return null;
  const match = String(dateStr).match(/\b(19\d{2}|20\d{2})\b/);
  if (match) {
    return parseInt(match[1], 10);
  }
  return null;
}

function computeDecade(year) {
  if (!year || year < 1900 || year > 2099) return null;
  return `${Math.floor(year / 10) * 10}s`;
}

function parseFilenameFallback(filename) {
  let stem = path.parse(filename).name;
  stem = stem.replace(/^\d+[\s\.\-_]+/, "");
  if (stem.includes(" - ")) {
    const parts = stem.split(" - ");
    const artist = parts[0].trim();
    const title = parts.slice(1).join(" - ").trim();
    return { artist: artist || null, title: title || stem };
  }
  return { artist: null, title: stem };
}

async function readAudioFile(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  if (!SUPPORTED_EXTENSIONS.has(ext)) return null;

  let stats;
  try {
    stats = fs.statSync(filePath);
    if (!stats.isFile()) return null;
  } catch (_) {
    return null;
  }

  const fileName = path.basename(filePath);
  const fileSize = stats.size;
  let duration = 0.0;

  let title = null;
  let artist = null;
  let album = null;
  let year = null;
  let parentGenre = null;
  let genre = null;
  let subGenre = null;
  let tempo = null;
  let hasCover = 0;

  try {
    const meta = await mm.parseFile(filePath, { duration: true, skipCovers: false });
    if (meta && meta.format && meta.format.duration) {
      duration = Math.round(meta.format.duration * 10) / 10;
    }

    if (meta && meta.common) {
      const c = meta.common;
      title = cleanTag(c.title);
      artist = cleanTag(c.artist);
      album = cleanTag(c.album);

      if (c.year) {
        year = parseYear(c.year);
      } else if (c.date) {
        year = parseYear(c.date);
      }

      if (c.picture && c.picture.length > 0) {
        hasCover = 1;
      }

      if (c.bpm) {
        tempo = `~${c.bpm} BPM`;
      }

      // Check genre
      if (Array.isArray(c.genre) && c.genre.length > 0) {
        const fullGenreStr = c.genre.join("; ");
        if (fullGenreStr.includes(";")) {
          const parts = fullGenreStr.split(";").map(p => p.trim()).filter(Boolean);
          if (parts.length >= 3) {
            parentGenre = parts[0];
            genre = parts[1];
            subGenre = parts[2];
          } else if (parts.length === 2) {
            parentGenre = parts[0];
            genre = parts[1];
          } else if (parts.length === 1) {
            genre = parts[0];
          }
        } else {
          genre = cleanTag(c.genre[0]);
        }
      }
    }

    // Check native custom frames for TXXX:PARENT_GENRE, TXXX:GENRE, TXXX:SUBGENRE, TXXX:TEMPO
    if (meta && meta.native) {
      const allNative = [];
      Object.values(meta.native).forEach(tagList => {
        if (Array.isArray(tagList)) allNative.push(...tagList);
      });

      allNative.forEach(tag => {
        const id = (tag.id || "").toUpperCase();
        const val = cleanTag(tag.value);
        if (!val) return;

        if (id === "TXXX:PARENT_GENRE" || id === "PARENT_GENRE" || id === "----:COM.APPLE.ITUNES:PARENT_GENRE") {
          parentGenre = val;
        } else if (id === "TXXX:GENRE" || id === "GENRE") {
          genre = val;
        } else if (id === "TXXX:SUBGENRE" || id === "TXXX:SUB_GENRE" || id === "SUB_GENRE" || id === "SUBGENRE") {
          subGenre = val;
        } else if (id === "TXXX:TEMPO" || id === "TXXX:PACE" || id === "TEMPO" || id === "PACE" || id === "----:COM.APPLE.ITUNES:TEMPO") {
          tempo = val;
        }
      });
    }
  } catch (err) {
    // If metadata parsing fails, fallback to filename parsing
  }

  // Fallback for title/artist from filename
  if (!title || !artist) {
    const fb = parseFilenameFallback(fileName);
    if (!title) title = fb.title;
    if (!artist) artist = fb.artist;
  }

  const decade = computeDecade(year);
  if (!parentGenre && genre) {
    parentGenre = inferParentFromGenre(genre);
  }

  const tagsList = formatTrackTags(parentGenre, genre, subGenre, decade, []);

  return {
    file_path: filePath,
    file_name: fileName,
    file_size: fileSize,
    duration,
    title,
    artist,
    album,
    release_year: year,
    decade,
    parent_genre: parentGenre,
    genre,
    sub_genre: subGenre,
    tempo,
    tags: tagsList,
    has_cover: hasCover,
    status: "unprocessed"
  };
}

async function scanDirectory(directoryPath) {
  const results = [];

  async function walk(dir) {
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch (_) {
      return;
    }

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        await walk(fullPath);
      } else if (entry.isFile()) {
        const ext = path.extname(entry.name).toLowerCase();
        if (SUPPORTED_EXTENSIONS.has(ext)) {
          const track = await readAudioFile(fullPath);
          if (track) {
            results.push(track);
          }
        }
      }
    }
  }

  await walk(directoryPath);
  return results;
}

// In-memory cover art cache
const coverCache = new Map();
const MAX_COVER_CACHE = 2000;

async function extractCoverBytes(filePath) {
  if (coverCache.has(filePath)) {
    return coverCache.get(filePath);
  }

  try {
    const meta = await mm.parseFile(filePath, { duration: false, skipCovers: false });
    if (meta && meta.common && meta.common.picture && meta.common.picture.length > 0) {
      const pic = meta.common.picture[0];
      const coverData = {
        data: Buffer.from(pic.data),
        mime: pic.format || "image/jpeg"
      };

      if (coverCache.size >= MAX_COVER_CACHE) {
        // Evict oldest 20%
        const keys = Array.from(coverCache.keys()).slice(0, Math.floor(MAX_COVER_CACHE * 0.2));
        keys.forEach(k => coverCache.delete(k));
      }

      coverCache.set(filePath, coverData);
      return coverData;
    }
  } catch (err) {
    console.error(`Error extracting cover from ${filePath}:`, err.message);
  }

  return null;
}

function invalidateCoverCache(filePath) {
  coverCache.delete(filePath);
}

module.exports = {
  SUPPORTED_EXTENSIONS,
  readAudioFile,
  scanDirectory,
  extractCoverBytes,
  invalidateCoverCache
};
